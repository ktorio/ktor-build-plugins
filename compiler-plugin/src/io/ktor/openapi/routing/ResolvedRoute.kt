package io.ktor.openapi.routing

data class ResolvedRoute(
    val path: String,
    val method: String,
    val fields: RouteFieldList = emptyList(),
)