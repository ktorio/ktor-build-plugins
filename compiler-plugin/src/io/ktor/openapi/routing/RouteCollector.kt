package io.ktor.openapi.routing

import org.jetbrains.kotlin.fir.FirSession

object RouteCollector {
    fun collectRoutes(graph: RouteCallGraph): List<ResolvedRoute> {
        // Process each route node to generate ResolvedRoutes
        return graph.routes.flatMap { route ->
            if (route.method == null) return@flatMap emptyList()
            
            // Find all paths from roots to this route
            val allPaths = graph.findAllPathsToRoots(route)
            
            // Process each path and merge fields
            with(graph.session) {
                allPaths.mapNotNull { path ->
                    resolve(
                        parents = path,
                        children = graph.getChildren(route)
                    )
                }
            }
        }
    }

    context(session: FirSession)
    fun resolve(
        parents: Iterable<RouteNode> = emptyList(),
        children: Iterable<RouteNode> = emptyList(),
    ): ResolvedRoute? {
        var stack = RouteStack(session)
        var fields = emptyList<RouteField>()
        for (ancestor in parents) {
            fields = ancestor.fields(stack).merge(fields)
            stack += ancestor
        }

        for (child in children) {
            fields = fields.merge(child.fields(stack))
        }

        val path = fields.path ?: return null
        val method = fields.method ?: return null
        if (RouteField.Ignore in fields) return null

        return ResolvedRoute(
            path = path,
            method = method,
            fields = fields,
        )
    }
}