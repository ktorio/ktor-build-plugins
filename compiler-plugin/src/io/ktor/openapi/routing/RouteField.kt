package io.ktor.openapi.routing

import io.ktor.openapi.model.JsonSchema
import io.ktor.openapi.model.JsonType
import kotlinx.serialization.json.JsonElement
import org.jetbrains.kotlin.utils.addToStdlib.firstIsInstanceOrNull

/**
 * Sealed class representing different KDoc parameters for OpenAPI documentation.
 */
sealed interface RouteField {

    fun merge(other: RouteField): RouteField? =
        if (this::class == other::class) this else null

    sealed interface SchemaHolder : RouteField {
        val schema: SchemaReference?
        val attributes: Map<String, JsonElement>
    }

    sealed interface Content : SchemaHolder {
        val contentType: String?
        val description: String?
    }

    sealed interface Parameter : SchemaHolder {
        val name: String
        val description: String?
        val `in`: String
    }

    /**
     * Built into the OpenAPI JSON structure, so ignored when building the path item.
     */
    sealed interface Transient : RouteField

    data class Method(val method: String) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Method && method == other.method) this else null
    }

    data class Path(val path: String) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Path)
                Path("${other.path}/${path.removePrefix("/")}")
            else null
    }

    data class Schema(
        val name: String,
        val schema: JsonSchema
    ) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Schema && other.name == name) this else null
    }

    /**
     * Associates the endpoint with a tag for grouping.
     *
     * Format: `@tag TagName`
     */
    data class Tag(val name: String) : RouteField {
        override fun merge(other: RouteField): RouteField? =
            if (other is Tag && name == other.name) this else null
    }

    /**
     * Describes a path parameter.
     *
     * Format: `@path name description`
     */
    data class PathParam(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "path"
        override fun merge(other: RouteField): RouteField? =
            if (other is PathParam && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Describes a query parameter.
     *
     * Format: `@query name description`
     */
    data class QueryParam(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "query"
        override fun merge(other: RouteField): RouteField? =
            if (other is QueryParam && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Describes a header parameter.
     *
     * Format: `@header name description`
     */
    data class Header(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "header"
        override fun merge(other: RouteField): RouteField? =
            if (other is Header && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
            ) else null
    }

    /**
     * Describes a cookie parameter.
     *
     * Format: `@cookie name description`
     */
    data class Cookie(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "cookie"
        override fun merge(other: RouteField): RouteField? =
            if (other is Cookie && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Documents the request body type.
     *
     * Format: `@body [Type] description`
     */
    data class Body(
        override val contentType: String? = null,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Content {
        override fun merge(other: RouteField): RouteField? =
            if (other is Body) copy(
                contentType = contentType ?: other.contentType,
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Documents a response code with optional type and description.
     *
     * Format: `@response code [Type] description`
     */
    data class Response(
        val code: String,
        override val contentType: String? = null,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Content {
        override fun merge(other: RouteField): RouteField? =
            if (other is Response && code == other.code)
                copy(
                    contentType = contentType ?: other.contentType,
                    schema = schema ?: other.schema,
                    description = description ?: other.description,
                    attributes = attributes + other.attributes,
                )
            else null
    }

    /**
     * Marks an endpoint as deprecated.
     *
     * Format: `@deprecated reason`
     */
    data class Deprecated(val reason: String) : RouteField

    /**
     * Provides a detailed endpoint description.
     *
     * Format: `@description text`
     */
    data class Description(val text: String) : RouteField


    /**
     * Provides a summary of the endpoint.
     * Format: `@summary text`
     */
    data class Summary(val text: String) : RouteField

    /**
     * Documents security requirements.
     *
     * Format: `@security scheme`
     */
    data class Security(val scheme: String) : RouteField {
        override fun merge(other: RouteField): RouteField? =
            if (other is Security && scheme == other.scheme) this else null
    }
}

typealias RouteFieldList = List<RouteField>

val RouteFieldList.path: String?
    get() = firstIsInstanceOrNull<RouteField.Path>()?.path

val RouteFieldList.method: String?
    get() = firstIsInstanceOrNull<RouteField.Method>()?.method

/**
 * Merges two lists of route fields into a new list.
 */
fun RouteFieldList.merge(other: RouteFieldList) = buildList {
    val otherMutable = other.toMutableList()
    for (field in this@merge) {
        val match = otherMutable.indices.firstNotNullOfOrNull { i ->
            field.merge(otherMutable[i])?.also {
                otherMutable.removeAt(i)
                add(it)
            }
        }
        // if no compatible field was found, add this field to the result
        if (match == null) add(field)
    }
    // add all remaining unmerged fields to the result
    // TODO ignore all summary fields because we currently have some false positives
    addAll(otherMutable.filterNot { it is RouteField.Summary })
}

sealed interface SchemaReference {

    fun asSchema(): JsonSchema

    data class Resolved(
        val schema: JsonSchema
    ) : SchemaReference {
        override fun asSchema(): JsonSchema = schema
    }

    sealed interface Link: SchemaReference {
        val name: String

        data class Simple(override val name: String, val jsonType: JsonType) : Link {
            override fun asSchema(): JsonSchema =
                JsonSchema(jsonType)
        }
        data class Reference(override val name: String) : Link {
            override fun asSchema(): JsonSchema =
                JsonSchema(ref = "#/components/schemas/$name")
        }
        data class Array(val element: Link) : Link by element {
            override fun asSchema(): JsonSchema =
                JsonSchema(
                    type = JsonType.array,
                    items = element.asSchema()
                )
        }
        data class Optional(val delegate: Link) : Link by delegate {
            override fun asSchema(): JsonSchema =
                delegate.asSchema()
                    .copy(required = false)
        }
    }
}

fun SchemaReference.hasReference(): Boolean = getReference() != null

fun SchemaReference.getReference(): String? = when(this) {
    is SchemaReference.Resolved -> schema.ref?.substringAfterLast('/')
    is SchemaReference.Link.Array -> element.getReference()
    is SchemaReference.Link.Optional -> delegate.getReference()
    is SchemaReference.Link.Reference -> name
    is SchemaReference.Link.Simple -> null
}