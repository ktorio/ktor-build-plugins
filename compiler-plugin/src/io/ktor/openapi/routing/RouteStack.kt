package io.ktor.openapi.routing

import io.ktor.compiler.utils.*
import org.jetbrains.kotlin.fir.FirSession
import org.jetbrains.kotlin.fir.expressions.FirExpression

data class RouteStack(
    val session: FirSession,
    val stack: List<RouteNode> = emptyList(),
) {
    operator fun plus(node: RouteNode): RouteStack =
        RouteStack(session, stack + node)

    fun FirExpression.evaluate() =
        with(session) {
            createCallStackEvaluator(stack.map { it.fir })
        }.evaluate(this)
}