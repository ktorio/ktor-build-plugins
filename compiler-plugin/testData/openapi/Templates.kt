// RUN_PIPELINE_TILL: BACKEND

package openapi

import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import io.ktor.server.application.*
import io.ktor.server.plugins.contentnegotiation.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

const val API_ROOT = "/api/v1"
const val API_VERSION = "v1"
const val API_LATEST = "latest"
const val ID_PARAM = "id"
const val UUID_PARAM = "uuid"
const val NAME_PARAM = "name"
const val QUERY_PARAM = "q"
const val SORT_PARAM = "sort"
const val PAGE_PARAM = "page"
const val SIZE_PARAM = "size"
const val FILTER_PARAM = "filter"

object EntityType {
    const val REPORTS = "reports"
    const val USERS = "users"
    const val ITEMS = "items"
    const val ANALYTICS = "analytics"
    const val DOCUMENTS = "documents"
}

const val UUID_REGEX = "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
const val DATE_REGEX = "\\d{4}-\\d{2}-\\d{2}"

fun Application.templateRoutes(
    reportService: ReportService,
    userService: UserService,
    itemService: ItemService,
    analyticsService: AnalyticsService,
    documentService: DocumentService
) {
    install(ContentNegotiation) {
        json()
    }

    routing {
        route(API_ROOT) {
            /**
             * Get all reports.
             * @response 200 [List<Report>] List of reports
             */
            get("/${EntityType.REPORTS}") {
                call.respond(reportService.getAllReports())
            }

            /**
             * Get report by ID.
             * @path id [Long] Report ID
             *   minimum: 1
             *   maximum: 999999999999
             *   required: true
             * @response 200 [Report] Report details
             * @response 404 Report not found
             */
            get("/${EntityType.REPORTS}/{$ID_PARAM}") {
                val id = call.parameters[ID_PARAM]?.toLongOrNull()
                    ?: return@get call.respond(HttpStatusCode.BadRequest, "Invalid ID format")

                val report = reportService.getReportById(id)
                if (report != null) {
                    call.respond(report)
                } else {
                    call.respond(HttpStatusCode.NotFound, "Report not found")
                }
            }

            /**
             * Get user reports.
             * @path userId [Long] User ID
             * @response 200 [List<Report>] User reports
             */
            get("/${EntityType.USERS}/{userId}/${EntityType.REPORTS}") {
                val userId = call.parameters["userId"]?.toLongOrNull()
                    ?: return@get call.respond(HttpStatusCode.BadRequest, "Invalid user ID")

                call.respond(reportService.getReportsByUserId(userId))
            }

            /**
             * Get filtered user reports for a specific date.
             * @path userId [Long] User ID
             * @path date [String] Report date (YYYY-MM-DD)
             *   pattern: $DATE_REGEX
             * @query type [String] Report type
             * @response 200 [List<Report>] Filtered user reports
             */
            get("/${EntityType.USERS}/{userId}/${EntityType.REPORTS}/{date}") {
                val userId = call.parameters["userId"]?.toLongOrNull()
                    ?: return@get call.respond(HttpStatusCode.BadRequest, "Invalid user ID")

                val date = call.parameters["date"]
                if (date == null || !date.matches(Regex(DATE_REGEX))) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid date format, use YYYY-MM-DD")
                }

                val type = call.request.queryParameters["type"]
                call.respond(reportService.getFilteredReportsByUserIdAndDate(userId, date, type))
            }
        }

        /**
         * Get API information for specific version.
         * @path version [String] API version
         *   enum: [v1, v2, latest]
         * @response 200 [ApiInfo] API information
         */
        get("/api/{version}/info") {
            val version = call.parameters["version"] ?: API_LATEST
            call.respond(ApiInfo(version = version, status = "active"))
        }

        /**
         * Get item by UUID.
         * @path uuid [String] Item UUID
         *   pattern: $UUID_REGEX
         * @response 200 [Item] Item details
         * @response 404 Item not found
         */
        get("/items/{$UUID_PARAM}") {
            val uuid = call.parameters[UUID_PARAM]
            if (uuid == null || !uuid.matches(Regex(UUID_REGEX))) {
                return@get call.respond(HttpStatusCode.BadRequest, "Invalid UUID format")
            }

            val item = itemService.getItemByUuid(uuid)
            if (item != null) {
                call.respond(item)
            } else {
                call.respond(HttpStatusCode.NotFound, "Item not found")
            }
        }

        /**
         * Get analytics data.
         * @path type [String] Analytics type
         *   enum: [daily, weekly, monthly, yearly]
         * @path metric [String] Metric name
         * @response 200 [AnalyticsData] Analytics data
         */
        get("/analytics/{type}/{metric}") {
            val type = call.parameters["type"] ?: "daily"
            val metric = call.parameters["metric"] ?: "views"

            call.respond(analyticsService.getAnalyticsData(type, metric))
        }

        route("/documents") {
            /**
             * Search documents.
             * @query q [String] Search query
             * @query type [String] Document type
             * @response 200 [List<Document>] Search results
             */
            get {
                val query = call.request.queryParameters[QUERY_PARAM] ?: ""
                val type = call.request.queryParameters["type"]

                call.respond(documentService.searchDocuments(query, type))
            }

            /**
             * Get document by ID.
             * @path id [String] Document ID
             * @response 200 [Document] Document details
             * @response 404 Document not found
             */
            get("/{$ID_PARAM}") {
                val id = call.parameters[ID_PARAM] ?: return@get call.respond(HttpStatusCode.BadRequest, "Missing ID")

                val document = documentService.getDocumentById(id)
                if (document != null) {
                    call.respond(document)
                } else {
                    call.respond(HttpStatusCode.NotFound, "Document not found")
                }
            }

            /**
             * Get document content.
             * @path id [String] Document ID
             * @path format [String] Content format
             *   enum: [html, text, pdf]
             * @response 200 Document content
             * @response 404 Document not found
             */
            get("/{$ID_PARAM}/content/{format}") {
                val id = call.parameters[ID_PARAM] ?: return@get call.respond(HttpStatusCode.BadRequest, "Missing ID")
                val format = call.parameters["format"] ?: "text"

                val contentType = when (format) {
                    "html" -> ContentType.Text.Html
                    "pdf" -> ContentType.Application.Pdf
                    else -> ContentType.Text.Plain
                }

                val content = documentService.getDocumentContent(id, format)
                if (content != null) {
                    call.respondText(content, contentType)
                } else {
                    call.respond(HttpStatusCode.NotFound, "Document not found")
                }
            }
        }

        route("/${EntityType.USERS}") {
            /**
             * Get all users.
             * @query sort [String] Sort field
             *   enum: [name, created, role]
             * @query order [String] Sort order
             *   enum: [asc, desc]
             * @response 200 [List<UserProfileTemplate>] List of users
             */
            get {
                val sort = call.request.queryParameters[SORT_PARAM] ?: "name"
                val order = call.request.queryParameters["order"] ?: "asc"

                call.respond(userService.getAllUsers(sort, order))
            }

            /**
             * Get user by ID.
             * @path id [Long] User ID
             * @response 200 [UserProfileTemplate] User details
             * @response 404 User not found
             */
            get("/{$ID_PARAM}") {
                val id = call.parameters[ID_PARAM]?.toLongOrNull()
                    ?: return@get call.respond(HttpStatusCode.BadRequest, "Invalid ID format")

                val user = userService.getUserById(id)
                if (user != null) {
                    call.respond(user)
                } else {
                    call.respond(HttpStatusCode.NotFound, "User not found")
                }
            }

            /**
             * Get user items.
             * @path id [Long] User ID
             * @query filter [String] Item filter
             * @response 200 [List<Item>] User items
             */
            get("/{$ID_PARAM}/${EntityType.ITEMS}") {
                val id = call.parameters[ID_PARAM]?.toLongOrNull()
                    ?: return@get call.respond(HttpStatusCode.BadRequest, "Invalid ID format")

                val filter = call.request.queryParameters[FILTER_PARAM]
                call.respond(itemService.getItemsByUserId(id, filter))
            }
        }

        /**
         * Get custom report.
         * @path year [Int] Report year
         *   minimum: 2000
         *   maximum: 2030
         * @path month [Int] Report month
         *   minimum: 1
         *   maximum: 12
         * @path day [Int] Report day
         *   minimum: 1
         *   maximum: 31
         * @path format [String] Report format
         *   enum: [json, csv, xml]
         * @response 200 Custom report
         */
        get("/reports/{year}/{month}/{day}.{format}") {
            val year = call.parameters["year"]?.toIntOrNull() ?: 2023
            val month = call.parameters["month"]?.toIntOrNull() ?: 1
            val day = call.parameters["day"]?.toIntOrNull() ?: 1
            val format = call.parameters["format"] ?: "json"

            val contentType = when (format) {
                "csv" -> ContentType.Text.CSV
                "xml" -> ContentType.Application.Xml
                else -> ContentType.Application.Json
            }

            val report = reportService.getCustomReport(year, month, day, format)
            call.respondText(report, contentType)
        }
    }
}

interface ReportService {
    fun getAllReports(): List<Report>
    fun getReportById(id: Long): Report?
    fun getReportsByUserId(userId: Long): List<Report>
    fun getFilteredReportsByUserIdAndDate(userId: Long, date: String, type: String?): List<Report>
    fun getCustomReport(year: Int, month: Int, day: Int, format: String): String
}

interface UserService {
    fun getAllUsers(sort: String, order: String): List<UserProfileTemplate>
    fun getUserById(id: Long): UserProfileTemplate?
}

interface ItemService {
    fun getItemByUuid(uuid: String): Item?
    fun getItemsByUserId(userId: Long, filter: String?): List<Item>
}

interface AnalyticsService {
    fun getAnalyticsData(type: String, metric: String): AnalyticsData
}

interface DocumentService {
    fun searchDocuments(query: String, type: String?): List<Document>
    fun getDocumentById(id: String): Document?
    fun getDocumentContent(id: String, format: String): String?
}

@Serializable
data class Report(
    val id: Long,
    val title: String,
    val type: String,
    val date: String,
    val data: Map<String, Double>
)

@Serializable
data class UserProfileTemplate(
    val id: Long,
    val username: String,
    val email: String,
    val role: String
)

@Serializable
data class Item(
    val uuid: String,
    val name: String,
    val description: String,
    val price: Double
)

@Serializable
data class AnalyticsData(
    val type: String,
    val metric: String,
    val values: List<DataPoint>
)

@Serializable
data class DataPoint(
    val label: String,
    val value: Double
)

@Serializable
data class Document(
    val id: String,
    val title: String,
    val type: String,
    val createdAt: String
)

@Serializable
data class ApiInfo(
    val version: String,
    val status: String
)