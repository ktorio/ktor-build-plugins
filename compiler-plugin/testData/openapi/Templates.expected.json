{
    "openapi": "3.1.1",
    "info": {
        "title": "Open API Document",
        "version": "1.0.0"
    },
    "paths": {
        "/api/v1/reports": {
            "get": {
                "summary": "Get all reports.",
                "responses": {
                    "200": {
                        "description": "List of reports",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/List<Report>"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/api/v1/reports/{id}": {
            "get": {
                "summary": "Get report by ID.",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "Report ID",
                        "required": true,
                        "schema": {
                            "type": "integer",
                            "minimum": 1,
                            "maximum": 999999999999,
                            "required": true
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Report details",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/Report"
                                }
                            }
                        }
                    },
                    "404": {
                        "description": "Report not found",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "400": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/api/v1/users/{userId}/reports": {
            "get": {
                "summary": "Get user reports.",
                "parameters": [
                    {
                        "name": "userId",
                        "in": "path",
                        "description": "User ID",
                        "required": true,
                        "schema": {
                            "type": "integer"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "User reports",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/List<Report>"
                                }
                            }
                        }
                    },
                    "400": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/api/v1/users/{userId}/reports/{date}": {
            "get": {
                "summary": "Get filtered user reports for a specific date.",
                "parameters": [
                    {
                        "name": "userId",
                        "in": "path",
                        "description": "User ID",
                        "required": true,
                        "schema": {
                            "type": "integer"
                        }
                    },
                    {
                        "name": "date",
                        "in": "path",
                        "description": "Report date (YYYY-MM-DD)",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "pattern": "$DATE_REGEX"
                        }
                    },
                    {
                        "name": "type",
                        "in": "query",
                        "description": "Report type",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Filtered user reports",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/List<Report>"
                                }
                            }
                        }
                    },
                    "400": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/api/{version}/info": {
            "get": {
                "summary": "Get API information for specific version.",
                "parameters": [
                    {
                        "name": "version",
                        "in": "path",
                        "description": "API version",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "enum": "[v1, v2, latest]"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "API information",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/ApiInfo"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/items/{uuid}": {
            "get": {
                "summary": "Get item by UUID.",
                "parameters": [
                    {
                        "name": "uuid",
                        "in": "path",
                        "description": "Item UUID",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "pattern": "$UUID_REGEX"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Item details",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/Item"
                                }
                            }
                        }
                    },
                    "404": {
                        "description": "Item not found",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "400": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/analytics/{type}/{metric}": {
            "get": {
                "summary": "Get analytics data.",
                "parameters": [
                    {
                        "name": "type",
                        "in": "path",
                        "description": "Analytics type",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "enum": "[daily, weekly, monthly, yearly]"
                        }
                    },
                    {
                        "name": "metric",
                        "in": "path",
                        "description": "Metric name",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Analytics data",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/AnalyticsData"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/documents": {
            "get": {
                "summary": "Search documents.",
                "parameters": [
                    {
                        "name": "q",
                        "in": "query",
                        "description": "Search query",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "type",
                        "in": "query",
                        "description": "Document type",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Search results",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/List<Document>"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/documents/{id}": {
            "get": {
                "summary": "Get document by ID.",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "Document ID",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Document details",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/Document"
                                }
                            }
                        }
                    },
                    "404": {
                        "description": "Document not found",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "400": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/documents/{id}/content/{format}": {
            "get": {
                "summary": "Get document content.",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "Document ID",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "format",
                        "in": "path",
                        "description": "Content format",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "enum": "[html, text, pdf]"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Document content",
                        "content": {
                            "contenttype": {}
                        }
                    },
                    "404": {
                        "description": "Document not found",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "400": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/users": {
            "get": {
                "summary": "Get all users.",
                "parameters": [
                    {
                        "name": "sort",
                        "in": "query",
                        "description": "Sort field",
                        "required": false,
                        "schema": {
                            "type": "string",
                            "enum": "[name, created, role]"
                        }
                    },
                    {
                        "name": "order",
                        "in": "query",
                        "description": "Sort order",
                        "required": false,
                        "schema": {
                            "type": "string",
                            "enum": "[asc, desc]"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "List of users",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/List<UserProfileTemplate>"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/users/{id}": {
            "get": {
                "summary": "Get user by ID.",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "User ID",
                        "required": true,
                        "schema": {
                            "type": "integer"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "User details",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/UserProfileTemplate"
                                }
                            }
                        }
                    },
                    "404": {
                        "description": "User not found",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    },
                    "400": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/users/{id}/items": {
            "get": {
                "summary": "Get user items.",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "User ID",
                        "required": true,
                        "schema": {
                            "type": "integer"
                        }
                    },
                    {
                        "name": "filter",
                        "in": "query",
                        "description": "Item filter",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "User items",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/List<Item>"
                                }
                            }
                        }
                    },
                    "400": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/reports/{year}/{month}/{day}.{format}": {
            "get": {
                "summary": "Get custom report.",
                "parameters": [
                    {
                        "name": "year",
                        "in": "path",
                        "description": "Report year",
                        "required": true,
                        "schema": {
                            "type": "integer",
                            "minimum": 2000,
                            "maximum": 2030
                        }
                    },
                    {
                        "name": "month",
                        "in": "path",
                        "description": "Report month",
                        "required": true,
                        "schema": {
                            "type": "integer",
                            "minimum": 1,
                            "maximum": 12
                        }
                    },
                    {
                        "name": "day",
                        "in": "path",
                        "description": "Report day",
                        "required": true,
                        "schema": {
                            "type": "integer",
                            "minimum": 1,
                            "maximum": 31
                        }
                    },
                    {
                        "name": "format",
                        "in": "path",
                        "description": "Report format",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "enum": "[json, csv, xml]"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Custom report",
                        "content": {
                            "contenttype": {}
                        }
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {
            "Report": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer"
                    },
                    "title": {
                        "type": "string"
                    },
                    "type": {
                        "type": "string"
                    },
                    "date": {
                        "type": "string"
                    },
                    "data": {
                        "type": "object",
                        "additionalProperties": {
                            "type": "number"
                        }
                    }
                }
            },
            "ApiInfo": {
                "type": "object",
                "properties": {
                    "version": {
                        "type": "string"
                    },
                    "status": {
                        "type": "string"
                    }
                }
            },
            "Item": {
                "type": "object",
                "properties": {
                    "uuid": {
                        "type": "string"
                    },
                    "name": {
                        "type": "string"
                    },
                    "description": {
                        "type": "string"
                    },
                    "price": {
                        "type": "number"
                    }
                }
            },
            "AnalyticsData": {
                "type": "object",
                "properties": {
                    "type": {
                        "type": "string"
                    },
                    "metric": {
                        "type": "string"
                    },
                    "values": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "label": {
                                    "type": "string"
                                },
                                "value": {
                                    "type": "number"
                                }
                            }
                        }
                    }
                }
            },
            "Document": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "string"
                    },
                    "title": {
                        "type": "string"
                    },
                    "type": {
                        "type": "string"
                    },
                    "createdAt": {
                        "type": "string"
                    }
                }
            },
            "UserProfileTemplate": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer"
                    },
                    "username": {
                        "type": "string"
                    },
                    "email": {
                        "type": "string"
                    },
                    "role": {
                        "type": "string"
                    }
                }
            }
        }
    }
}